flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/factories/HiveFunctionDefinitionFactory.java

```diff
@@ -55,27 +57,56 @@ public HiveFunctionDefinitionFactory(HiveShim hiveShim) {
     @Override
     public FunctionDefinition createFunctionDefinition(
             String name, CatalogFunction catalogFunction) {
-        if (catalogFunction.isGeneric()) {
-            return createFunctionDefinitionFromFlinkFunction(name, catalogFunction);
+        return createFunctionDefinition(
+                name, catalogFunction, () -> Thread.currentThread().getContextClassLoader());
+    }
+
+    @Override
+    public FunctionDefinition createFunctionDefinition(
+            String name, CatalogFunction catalogFunction, Context context) {
+        if (isGenericFunction(catalogFunction, context.getClassLoader())) {
+            return createFunctionDefinitionFromFlinkFunction(name, catalogFunction, context);
         }
-        return createFunctionDefinitionFromHiveFunction(name, catalogFunction.getClassName());
+        return createFunctionDefinitionFromHiveFunction(
+                name, catalogFunction.getClassName(), context);
     }
 
     public FunctionDefinition createFunctionDefinitionFromFlinkFunction(
-            String name, CatalogFunction catalogFunction) {
+            String name, CatalogFunction catalogFunction, Context context) {
         return UserDefinedFunctionHelper.instantiateFunction(
-                Thread.currentThread().getContextClassLoader(), null, name, catalogFunction);
+                context.getClassLoader(), null, name, catalogFunction);
+    }
+
+    /**
+     * Distinguish if the function is a generic function.
+     *
+     * @return whether the function is a generic function
+     */
+    private boolean isGenericFunction(CatalogFunction catalogFunction, ClassLoader classLoader) {
```

Comment:
We can get rid of the `generic` term now, and call this method `isFlinkFunction`.